lib/instrumentation/modules/@aws-sdk/client-s3.js

```diff
@@ -0,0 +1,199 @@
+ */
+function s3MiddlewareFactory (client, agent) {
+  return [
+    {
+      middleware: (next, context) => async (args) => {
+        const input = args.input
+        const bucket = input && input.Bucket
+        const resource = resourceFromBucket(bucket)
+        const span = agent._instrumentation.currSpan()
+
+        if (!span) {
+          return await next(args)
+        }
+        // The given span comes with the operation name and we need to
+        // add the resource if applies
+        if (resource) {
+          span.name += ' ' + resource
+          span.setServiceTarget('s3', resource)
+        }
+
+        // As for now OTel spec defines attributes for operations that require a Bucket
+        // if that changes we should review this guard
+        // https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/semantic_conventions/trace/instrumentation/aws-sdk.yml#L435
+        if (bucket) {
+          const otelAttrs = span._getOTelAttributes()
+
+          otelAttrs['aws.s3.bucket'] = bucket
+
+          if (input.Key) {
+            otelAttrs['aws.s3.key'] = input.Key
+          }
+        }
+
+        let err
+        let result
+        let response
+        let statusCode
+        try {
+          result = await next(args)
+          response = result && result.response
+          statusCode = response && response.statusCode
+        } catch (ex) {
+          // Save the error for use in `finally` below, but re-throw it to
+          // not impact code flow.
+          err = ex
+
+          // This code path happens with a GetObject conditional request
+          // that returns a 304 Not Modified.
+          statusCode = err && err.$metadata && err.$metadata.httpStatusCode
+          throw ex
+        } finally {
+          if (statusCode) {
+            span._setOutcomeFromHttpStatusCode(statusCode)
+          } else {
+            span._setOutcomeFromErrorCapture(constants.OUTCOME_FAILURE)
+          }
+          if (err && (!statusCode || statusCode >= 400)) {
+            agent.captureError(err, { skipOutcome: true })
+          }
+
+          // Set the httpContext
+          if (statusCode) {
+            const httpContext = {
+              status_code: statusCode
+            }
+
+            if (response && response.headers) {
+              const encodedBodySize = Number(response.headers['content-length'])
+              if (encodedBodySize) {
+                httpContext.response = { encoded_body_size: encodedBodySize }
+              }
+            }
+            span.setHttpContext(httpContext)
+          }
+
+          // Configuring `new S3Client({useArnRegion:true})` allows one to
+          // use an Access Point bucket ARN for a region *other* than the
+          // one for which the client is configured. Therefore, we attempt
+          // to get the bucket region from the ARN first.
+          const useArnRegion = await client.config.useArnRegion()
```

Comment:
the config properties `useArnRegion` and `region` are defined in `ClientInputEndpointParamenters` interface. [See here](https://github.com/aws/aws-sdk-js-v3/blob/b1433c46aa476cd5574fb6d2942fb554f0451dbf/clients/client-s3/src/endpoint/EndpointParameters.ts#LL4C10-L4C10)

We can see the properties are defined as
- `useArnRegion` can be a boolean or a `Provider<boolean>`
- `region` can be a string or a `Provider<string>`

The generic type `Provider` defines a function that returns a promise of the given type so it can be awaited. 

The client configuration in the test suite always have these properties set to providers (default ones) but they can be defined at construction time passing a different [configuration object](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/s3clientconfig.html)

We need to check if those props are values or functions and act accordingly. Thanks for the heads up :)